weave-js/src/components/Panel2/PanelTable/util.ts

```diff
@@ -60,29 +62,60 @@ export const stripTag = (type: Type): Type => {
   return isTaggedValue(type) ? taggedValueValueType(type) : type;
 };
 
-// Very simple type shape comparison
+// Very simple type shape comparison util used to determine if tableState
+// needs to be reset or not.
 export const typeShapesMatch = (type: Type, toType: Type): boolean => {
+  function propertyTypesAreCompatible(
+    incomingTypedDict: TypedDictType,
+    toTypedDict: TypedDictType
+  ) {
+    for (const key of Object.keys(toTypedDict.propertyTypes)) {
+      const keyType = incomingTypedDict.propertyTypes[key];
+      const toKeyType = toTypedDict.propertyTypes[key];
+      if (keyType === undefined || !typeShapesMatch(keyType, toKeyType!)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
   type = stripTag(type);
   toType = stripTag(toType);
+
+  // List type handling
   if (isList(type) || isList(toType)) {
-    if (!isList(type) || !isList(toType)) {
-      return false;
-    } else {
-      return typeShapesMatch(listObjectType(type), listObjectType(toType));
+    return (
+      isList(type) &&
+      isList(toType) &&
+      typeShapesMatch(listObjectType(type), listObjectType(toType))
+    );
+  }
+  // TypedDict type handling
+  if (isTypedDict(type) || isTypedDict(toType)) {
+    // Handle union<typedDict> case
+    //
+    // This is meant for special cases where a user adds another column to their table
+    // in a subsequent run.
+    if (isTypedDict(toType) && isUnion(type)) {
+      return type.members.every(
+        member =>
+          isTypedDict(member) &&
+          isTypedDict(toType) &&
+          propertyTypesAreCompatible(member, toType)
+      );
     }
```

Comment:
everything else is just refactor (?), this lgtm

Reply:
Yup, besides the union checking code, I refactored to make it a bit more readable. Hope it didn't make the review hard to parse. Thanks for reviewing my PRs, @gtarpenning , I appreciate it! 🙏